There is a lint error: 
```
/noobaa-core/src/util/fips.js
  61:41  error  Unary word operator 'new' must be followed by whitespace  space-unary-ops
✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -131,7 +131,13 @@ module.exports = {
 
         'function-call-argument-newline': 'off',
 
-        'space-unary-ops': ['error', { words: false, nonwords: false }],
+        'space-unary-ops': ['error', {
+            words: true,
+            nonwords: false,
+            overrides: {
+                typeof: false,
+            },
+        }],
 
         // max line length is 80 by default, allow some slack
         // TODO eslint max-len for code lines should be error and reduced to ~100 instead of 140

--- a/src/endpoint/blob/blob_errors.js
+++ b/src/endpoint/blob/blob_errors.js
@@ -3,15 +3,25 @@
 
 const xml_utils = require('../../util/xml_utils');
 
+/**
+ * @typedef {{
+ *      code?: string, 
+ *      message: string, 
+ *      http_code: number,
+ *      detail?: string
+ * }} BlobErrorSpec
+ */
+
 class BlobError extends Error {
 
-    constructor(err) {
-        super(err.message);
-        this.code = err.reply_without_code ? undefined : err.code;
-        this.http_code = err.http_code;
-        if (err.reply) {
-            this.reply = err.reply;
-        }
+    /**
+     * @param {BlobErrorSpec} error_spec 
+     */
+    constructor({ code, message, http_code, detail }) {
+        super(message); // sets this.message
+        this.code = code;
+        this.http_code = http_code;
+        this.detail = detail;
     }
 
     reply() {
@@ -29,42 +39,45 @@ class BlobError extends Error {
  * https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes
  * https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-error-codes
  */
-const errors_defs = [{
+BlobError.InternalError = Object.freeze({
     code: 'InternalError',
     message: 'The server encountered an internal error. Please retry the request.',
     http_code: 500,
-}, {
+});
+BlobError.ContainerAlreadyExists = Object.freeze({
     code: 'ContainerAlreadyExists',
     message: 'The specified container already exists.',
     http_code: 409,
-}, {
+});
+BlobError.ContainerNotFound = Object.freeze({
     code: 'ContainerNotFound',
     message: 'The specified container does not exist.',
     http_code: 404,
-}, {
+});
+BlobError.NotImplemented = Object.freeze({
     code: 'NotImplemented',
     message: 'functionality not implemented.',
     http_code: 501,
-}, {
+});
+BlobError.InvalidBlobOrBlock = Object.freeze({
     code: 'InvalidBlobOrBlock',
     message: 'The specified blob or block content is invalid.',
     http_code: 400,
-}, {
+});
+BlobError.InvalidHeaderValue = Object.freeze({
     code: 'InvalidHeaderValue',
     message: 'The value provided for one of the HTTP headers was not in the correct format.',
     http_code: 400,
-}, {
+});
+BlobError.OutOfRangeQueryParameterValue = Object.freeze({
     code: 'OutOfRangeQueryParameterValue',
     message: 'A query parameter specified in the request URI is outside the permissible range.',
     http_code: 400,
-}, {
+});
+BlobError.BlobNotFound = Object.freeze({
     code: 'BlobNotFound',
     message: 'The specified blob does not exist.',
     http_code: 404,
-}];
-
-for (const err_def of errors_defs) {
-    BlobError[err_def.code] = err_def;
-}
+});
 
 exports.BlobError = BlobError;

--- a/src/endpoint/blob/blob_rest.js
+++ b/src/endpoint/blob/blob_rest.js
@@ -144,36 +144,36 @@ function handle_error(req, res, err) {
 }
 
 function load_ops() {
-    const r = x => require(x); // eslint-disable-line global-require
+    /* eslint-disable global-require */
     return js_utils.deep_freeze({
 
         // SERVICE
-        get_service_list: r('./ops/blob_get_service_list'),
-        get_service_stats: r('./ops/blob_get_service_stats'),
-        get_service_properties: r('./ops/blob_get_service_properties'),
-        put_service_properties: r('./ops/blob_put_service_properties'),
+        get_service_list: require('./ops/blob_get_service_list'),
+        get_service_stats: require('./ops/blob_get_service_stats'),
+        get_service_properties: require('./ops/blob_get_service_properties'),
+        put_service_properties: require('./ops/blob_put_service_properties'),
 
         // CONTAINER
-        get_container: r('./ops/blob_get_container'),
-        get_container_acl: r('./ops/blob_get_container_acl'),
-        get_container_list: r('./ops/blob_get_container_list'),
-        get_container_metadata: r('./ops/blob_get_container_metadata'),
-        put_container: r('./ops/blob_put_container'),
-        put_container_acl: r('./ops/blob_put_container_acl'),
-        put_container_lease: r('./ops/blob_put_container_lease'),
-        put_container_metadata: r('./ops/blob_put_container_metadata'),
-        delete_container: r('./ops/blob_delete_container'),
+        get_container: require('./ops/blob_get_container'),
+        get_container_acl: require('./ops/blob_get_container_acl'),
+        get_container_list: require('./ops/blob_get_container_list'),
+        get_container_metadata: require('./ops/blob_get_container_metadata'),
+        put_container: require('./ops/blob_put_container'),
+        put_container_acl: require('./ops/blob_put_container_acl'),
+        put_container_lease: require('./ops/blob_put_container_lease'),
+        put_container_metadata: require('./ops/blob_put_container_metadata'),
+        delete_container: require('./ops/blob_delete_container'),
 
         // BLOB
-        get_blob: r('./ops/blob_get_blob'),
-        get_blob_metadata: r('./ops/blob_get_blob_metadata'),
-        get_blob_blocklist: r('./ops/blob_get_blob_blocklist'),
-        put_blob: r('./ops/blob_put_blob'),
-        put_blob_block: r('./ops/blob_put_blob_block'),
-        put_blob_blocklist: r('./ops/blob_put_blob_blocklist'),
-        put_blob_properties: r('./ops/blob_put_blob_properties'),
-        put_blob_lease: r('./ops/blob_put_blob_lease'),
-        delete_blob: r('./ops/blob_delete_blob'),
+        get_blob: require('./ops/blob_get_blob'),
+        get_blob_metadata: require('./ops/blob_get_blob_metadata'),
+        get_blob_blocklist: require('./ops/blob_get_blob_blocklist'),
+        put_blob: require('./ops/blob_put_blob'),
+        put_blob_block: require('./ops/blob_put_blob_block'),
+        put_blob_blocklist: require('./ops/blob_put_blob_blocklist'),
+        put_blob_properties: require('./ops/blob_put_blob_properties'),
+        put_blob_lease: require('./ops/blob_put_blob_lease'),
+        delete_blob: require('./ops/blob_delete_blob'),
     });
 }
 

--- a/src/endpoint/lambda/lambda_errors.js
+++ b/src/endpoint/lambda/lambda_errors.js
@@ -1,15 +1,25 @@
 /* Copyright (C) 2016 NooBaa */
 'use strict';
 
+/**
+ * @typedef {{
+ *      code?: string, 
+ *      message: string, 
+ *      http_code: number,
+ *      detail?: string
+ * }} LambdaErrorSpec
+ */
+
 class LambdaError extends Error {
 
-    constructor(err) {
-        super(err.message);
-        this.code = err.reply_without_code ? undefined : err.code;
-        this.http_code = err.http_code;
-        if (err.reply) {
-            this.reply = err.reply;
-        }
+    /**
+     * @param {LambdaErrorSpec} error_spec 
+     */
+    constructor({ code, message, http_code, detail }) {
+        super(message); // sets this.message
+        this.code = code;
+        this.http_code = http_code;
+        this.detail = detail;
     }
 
     reply() {
@@ -20,84 +30,103 @@ class LambdaError extends Error {
 
 }
 
-const errors_defs = [{
+LambdaError.CodeStorageExceededException = Object.freeze({
     code: 'CodeStorageExceededException',
     message: 'You have exceeded your maximum total code size per account.',
     http_code: 400,
-}, {
+});
+LambdaError.InvalidParameterValueException = Object.freeze({
     code: 'InvalidParameterValueException',
     message: 'One of the parameters in the request is invalid.',
     http_code: 400,
-}, {
+});
+LambdaError.ResourceConflictException = Object.freeze({
     code: 'ResourceConflictException',
     message: 'The resource already exists.',
     http_code: 409,
-}, {
+});
+LambdaError.ResourceNotFoundException = Object.freeze({
     code: 'ResourceNotFoundException',
     message: 'The resource specified in the request does not exist.',
     http_code: 404,
-}, {
+});
+LambdaError.RequestTooLargeException = Object.freeze({
     code: 'RequestTooLargeException',
     message: 'The request payload exceeded the Invoke request body JSON input limit.',
     http_code: 413,
-}, {
+});
+LambdaError.ServiceException = Object.freeze({
     code: 'ServiceException',
     message: 'The AWS Lambda service encountered an internal error.',
     http_code: 500,
-}, {
+});
+LambdaError.TooManyRequestsException = Object.freeze({
     code: 'TooManyRequestsException',
     message: 'TooManyRequestsException.',
     http_code: 429,
-}, {
+});
+LambdaError.UnsupportedMediaTypeException = Object.freeze({
     code: 'UnsupportedMediaTypeException',
     message: 'The content type of the Invoke request body is not JSON.',
     http_code: 415,
-}, {
+});
+LambdaError.RequestTimeTooSkewed = Object.freeze({
     code: 'RequestTimeTooSkewed',
     message: 'The difference between the request time and the server\'s time is too large.',
     http_code: 403,
-}, {
+});
+LambdaError.AccessDenied = Object.freeze({
     code: 'AccessDenied',
     message: 'Access Denied',
     http_code: 403,
-}, {
+});
+LambdaError.SignatureDoesNotMatch = Object.freeze({
     code: 'SignatureDoesNotMatch',
     message: 'The request signature we calculated does not match the signature you provided. Check your AWS secret access key and signing method. For more information, see REST Authentication and SOAP Authentication for details.',
     http_code: 403,
-}, {
+});
+LambdaError.MissingRequestBodyError = Object.freeze({
     code: 'MissingRequestBodyError',
     message: 'Request body is empty.',
     http_code: 400,
-}, {
+});
+LambdaError.InvalidRequest = Object.freeze({
     code: 'InvalidRequest',
     message: 'The content type of the request is not JSON.',
     http_code: 400,
-}, {
+});
+LambdaError.InvalidDigest = Object.freeze({
     code: 'InvalidDigest',
     message: 'The Content-MD5 you specified is not valid.',
     http_code: 400,
-}, {
+});
+LambdaError.MaxMessageLengthExceeded = Object.freeze({
     code: 'MaxMessageLengthExceeded',
     message: 'Your request was too big.',
     http_code: 400,
-}, {
-    // not defined in AWS list
-    // needed for error cases that require no code in the reply
-    reply_without_code: true,
-    code: 'BadRequestWithoutCode',
+});
+
+////////////////////////////////////////////////////////////////
+// Errors actually returned by AWS S3 although not documented //
+////////////////////////////////////////////////////////////////
+
+LambdaError.NotImplemented = Object.freeze({
+    code: 'NotImplemented',
+    message: 'A header you provided implies functionality that is not implemented.',
+    http_code: 501,
+});
+LambdaError.BadRequestWithoutCode = Object.freeze({
+    // same as BadRequest but without encoding the <Code> field
+    // was needed for one of the cases in ceph/s3tests
     message: 'Bad Request',
     http_code: 400,
-}, {
-    // not defined in AWS docs but this is what they return
+});
+LambdaError.XAmzContentSHA256Mismatch = Object.freeze({
     code: 'XAmzContentSHA256Mismatch',
     message: 'The provided \'x-amz-content-sha256\' header does not match what was computed.',
     http_code: 400,
     // ClientComputedContentSHA256: '...',
     // S3ComputedContentSHA256: '...',
-}];
-
-for (const err_def of errors_defs) {
-    LambdaError[err_def.code] = err_def;
-}
+});
 
 exports.LambdaError = LambdaError;

--- a/src/endpoint/lambda/lambda_rest.js
+++ b/src/endpoint/lambda/lambda_rest.js
@@ -190,13 +190,13 @@ function handle_error(req, res, err) {
 }
 
 function load_ops() {
-    const r = x => require(x); // eslint-disable-line global-require
+    /* eslint-disable global-require */
     return js_utils.deep_freeze({
-        get_service: r('./ops/lambda_list_funcs'),
-        get_func: r('./ops/lambda_get_func'),
-        delete_func: r('./ops/lambda_delete_func'),
-        post_service: r('./ops/lambda_create_func'),
-        post_func_invocations: r('./ops/lambda_invoke_func'),
+        get_service: require('./ops/lambda_list_funcs'),
+        get_func: require('./ops/lambda_get_func'),
+        delete_func: require('./ops/lambda_delete_func'),
+        post_service: require('./ops/lambda_create_func'),
+        post_func_invocations: require('./ops/lambda_invoke_func'),
     });
 }
 

--- a/src/endpoint/s3/s3_errors.js
+++ b/src/endpoint/s3/s3_errors.js
@@ -3,16 +3,25 @@
 
 const xml_utils = require('../../util/xml_utils');
 
+/**
+ * @typedef {{
+ *      code?: string, 
+ *      message: string, 
+ *      http_code: number,
+ *      detail?: string
+ * }} S3ErrorSpec
+ */
+
 class S3Error extends Error {
 
-    constructor(err) {
-        super(err.message);
-        this.code = err.reply_without_code ? undefined : err.code;
-        this.http_code = err.http_code;
-        this.detail = err.detail;
-        if (err.reply) {
-            this.reply = err.reply;
-        }
+    /**
+     * @param {S3ErrorSpec} error_spec 
+     */
+    constructor({ code, message, http_code, detail }) {
+        super(message); // sets this.message
+        this.code = code;
+        this.http_code = http_code;
+        this.detail = detail;
     }
 
     reply(resource, request_id) {
@@ -21,12 +30,10 @@ class S3Error extends Error {
                 Code: this.code,
                 Message: this.message,
                 Resource: resource || '',
-                RequestId: request_id || ''
+                RequestId: request_id || '',
+                Detail: this.detail,
             }
         };
-        if (this.detail) {
-            xml.Error.Detail = this.detail;
-        }
         return xml_utils.encode_xml(xml);
     }
 
@@ -34,381 +41,465 @@ class S3Error extends Error {
 
 // See http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
 
-const errors_defs = [{
+
+S3Error.AccessDenied = Object.freeze({
     code: 'AccessDenied',
     message: 'Access Denied',
     http_code: 403,
-}, {
+});
+S3Error.AccountProblem = Object.freeze({
     code: 'AccountProblem',
     message: 'There is a problem with your AWS account that prevents the operation from completing successfully. Please Contact Us.',
     http_code: 403,
-}, {
+});
+S3Error.AmbiguousGrantByEmailAddress = Object.freeze({
     code: 'AmbiguousGrantByEmailAddress',
     message: 'The email address you provided is associated with more than one account.',
     http_code: 400,
-}, {
+});
+S3Error.BadDigest = Object.freeze({
     code: 'BadDigest',
     message: 'The Content-MD5 you specified did not match what we received.',
     http_code: 400,
-}, {
+});
+S3Error.BucketAlreadyExists = Object.freeze({
     code: 'BucketAlreadyExists',
     message: 'The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again.',
     http_code: 409,
-}, {
+});
+S3Error.BucketAlreadyOwnedByYou = Object.freeze({
     code: 'BucketAlreadyOwnedByYou',
     message: 'Your previous request to create the named bucket succeeded and you already own it. You get this error in all AWS regions except US East (N. Virginia) region, us-east-1. In us-east-1 region, you will get 200 OK, but it is no-op (if bucket exists it Amazon S3 will not do anything).',
     http_code: 409,
-}, {
+});
+S3Error.BucketNotEmpty = Object.freeze({
     code: 'BucketNotEmpty',
     message: 'The bucket you tried to delete is not empty. You must delete all versions in the bucket.',
     http_code: 409,
-}, {
+});
+S3Error.CredentialsNotSupported = Object.freeze({
     code: 'CredentialsNotSupported',
     message: 'This request does not support credentials.',
     http_code: 400,
-}, {
+});
+S3Error.CrossLocationLoggingProhibited = Object.freeze({
     code: 'CrossLocationLoggingProhibited',
     message: 'Cross-location logging not allowed. Buckets in one geographic location cannot log information to a bucket in another location.',
     http_code: 403,
-}, {
+});
+S3Error.EntityTooSmall = Object.freeze({
     code: 'EntityTooSmall',
     message: 'Your proposed upload is smaller than the minimum allowed object size.',
     http_code: 400,
-}, {
+});
+S3Error.EntityTooLarge = Object.freeze({
     code: 'EntityTooLarge',
     message: 'Your proposed upload exceeds the maximum allowed object size.',
     http_code: 400,
-}, {
+});
+S3Error.ExpiredToken = Object.freeze({
     code: 'ExpiredToken',
     message: 'The provided token has expired.',
     http_code: 400,
-}, {
+});
+S3Error.IllegalVersioningConfigurationException = Object.freeze({
     code: 'IllegalVersioningConfigurationException',
     message: 'Indicates that the versioning configuration specified in the request is invalid.',
     http_code: 400,
-}, {
+});
+S3Error.IncompleteBody = Object.freeze({
     code: 'IncompleteBody',
     message: 'You did not provide the number of bytes specified by the Content-Length HTTP header',
     http_code: 400,
-}, {
+});
+S3Error.IncorrectNumberOfFilesInPostRequest = Object.freeze({
     code: 'IncorrectNumberOfFilesInPostRequest',
     message: 'POST requires exactly one file upload per request.',
     http_code: 400,
-}, {
+});
+S3Error.InlineDataTooLarge = Object.freeze({
     code: 'InlineDataTooLarge',
     message: 'Inline data exceeds the maximum allowed size.',
     http_code: 400,
-}, {
+});
+S3Error.InternalError = Object.freeze({
     code: 'InternalError',
     message: 'We encountered an internal error. Please try again.',
     http_code: 500,
-}, {
+});
+S3Error.InvalidAccessKeyId = Object.freeze({
     code: 'InvalidAccessKeyId',
     message: 'The AWS access key Id you provided does not exist in our records.',
     http_code: 403,
-}, {
+});
+S3Error.InvalidAddressingHeader = Object.freeze({
     code: 'InvalidAddressingHeader',
     message: 'You must specify the Anonymous role.',
     http_code: 400, // N/A,
-}, {
+});
+S3Error.InvalidArgument = Object.freeze({
     code: 'InvalidArgument',
     message: 'Invalid Argument',
     http_code: 400,
-}, {
+});
+S3Error.InvalidBucketName = Object.freeze({
     code: 'InvalidBucketName',
     message: 'The specified bucket is not valid.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidBucketState = Object.freeze({
     code: 'InvalidBucketState',
     message: 'The request is not valid with the current state of the bucket.',
     http_code: 409,
-}, {
+});
+S3Error.InvalidDigest = Object.freeze({
     code: 'InvalidDigest',
     message: 'The Content-MD5 you specified is not valid.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidEncryptionAlgorithmError = Object.freeze({
     code: 'InvalidEncryptionAlgorithmError',
     message: 'The encryption request you specified is not valid. The valid value is AES256.',
     http_code: 400,
-}, {
+});
+S3Error.ServerSideEncryptionConfigurationNotFoundError = Object.freeze({
     code: 'ServerSideEncryptionConfigurationNotFoundError',
     message: 'The server side encryption configuration was not found.',
     http_code: 404,
-}, {
+});
+S3Error.InvalidLocationConstraint = Object.freeze({
     code: 'InvalidLocationConstraint',
     message: 'The specified location constraint is not valid. For more information about regions, see How to Select a Region for Your Buckets.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidObjectState = Object.freeze({
     code: 'InvalidObjectState',
     message: 'The operation is not valid for the current state of the object.',
     http_code: 403,
-}, {
+});
+S3Error.InvalidPart = Object.freeze({
     code: 'InvalidPart',
     message: 'One or more of the specified parts could not be found. The part might not have been uploaded, or the specified entity tag might not have matched the part\'s entity tag.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidPartOrder = Object.freeze({
     code: 'InvalidPartOrder',
     message: 'The list of parts was not in ascending order.Parts list must specified in order by part number.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidPayer = Object.freeze({
     code: 'InvalidPayer',
     message: 'All access to this object has been disabled.',
     http_code: 403,
-}, {
+});
+S3Error.InvalidPolicyDocument = Object.freeze({
     code: 'InvalidPolicyDocument',
     message: 'The content of the form does not meet the conditions specified in the policy document.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidRange = Object.freeze({
     code: 'InvalidRange',
     message: 'The requested range cannot be satisfied.',
     http_code: 416,
-}, {
+});
+S3Error.InvalidRequest = Object.freeze({
     code: 'InvalidRequest',
     message: 'SOAP requests must be made over an HTTPS connection.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidSecurity = Object.freeze({
     code: 'InvalidSecurity',
     message: 'The provided security credentials are not valid.',
     http_code: 403,
-}, {
+});
+S3Error.InvalidSOAPRequest = Object.freeze({
     code: 'InvalidSOAPRequest',
     message: 'The SOAP request body is invalid.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidStorageClass = Object.freeze({
     code: 'InvalidStorageClass',
     message: 'The storage class you specified is not valid.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidTargetBucketForLogging = Object.freeze({
     code: 'InvalidTargetBucketForLogging',
     message: 'The target bucket for logging does not exist, is not owned by you, or does not have the appropriate grants for the log-delivery group.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidToken = Object.freeze({
     code: 'InvalidToken',
     message: 'The provided token is malformed or otherwise invalid.',
     http_code: 400,
-}, {
+});
+S3Error.InvalidURI = Object.freeze({
     code: 'InvalidURI',
     message: 'Couldn\'t parse the specified URI.',
     http_code: 400,
-}, {
+});
+S3Error.KeyTooLong = Object.freeze({
     code: 'KeyTooLong',
     message: 'Your key is too long.',
     http_code: 400,
-}, {
+});
+S3Error.MalformedACLError = Object.freeze({
     code: 'MalformedACLError',
     message: 'The XML you provided was not well-formed or did not validate against our published schema.',
     http_code: 400,
-}, {
+});
+S3Error.MalformedPOSTRequest = Object.freeze({
     code: 'MalformedPOSTRequest',
     message: 'The body of your POST request is not well-formed multipart/form-data.',
     http_code: 400,
-}, {
+});
+S3Error.MalformedXML = Object.freeze({
     code: 'MalformedXML',
     message: 'This happens when the user sends malformed xml (xml that doesn\'t conform to the published xsd) for the configuration. The error message is, "The XML you provided was not well-formed or did not validate against our published schema."',
     http_code: 400,
-}, {
+});
+S3Error.InvalidTag = Object.freeze({
     code: 'InvalidTag',
     message: 'The tag provided was not a valid tag.',
     http_code: 400,
-}, {
+});
+S3Error.MaxMessageLengthExceeded = Object.freeze({
     code: 'MaxMessageLengthExceeded',
     message: 'Your request was too big.',
     http_code: 400,
-}, {
+});
+S3Error.MaxPostPreDataLengthExceededError = Object.freeze({
     code: 'MaxPostPreDataLengthExceededError',
     message: 'Your POST request fields preceding the upload file were too large.',
     http_code: 400,
-}, {
+});
+S3Error.MetadataTooLarge = Object.freeze({
     code: 'MetadataTooLarge',
     message: 'Your metadata headers exceed the maximum allowed metadata size.',
     http_code: 400,
-}, {
+});
+S3Error.MethodNotAllowed = Object.freeze({
     code: 'MethodNotAllowed',
     message: 'The specified method is not allowed against this resource.',
     http_code: 405,
-}, {
+});
+S3Error.MissingAttachment = Object.freeze({
     code: 'MissingAttachment',
     message: 'A SOAP attachment was expected, but none were found.',
     http_code: 400, // N/A,
-}, {
+});
+S3Error.MissingContentLength = Object.freeze({
     code: 'MissingContentLength',
     message: 'You must provide the Content-Length HTTP header.',
     http_code: 411,
-}, {
+});
+S3Error.MissingRequestBodyError = Object.freeze({
     code: 'MissingRequestBodyError',
     message: 'Request body is empty.',
     http_code: 400,
-}, {
+});
+S3Error.MissingSecurityElement = Object.freeze({
     code: 'MissingSecurityElement',
     message: 'The SOAP 1.1 request is missing a security element.',
     http_code: 400,
-}, {
+});
+S3Error.MissingSecurityHeader = Object.freeze({
     code: 'MissingSecurityHeader',
     message: 'Your request is missing a required header.',
     http_code: 400,
-}, {
+});
+S3Error.NoLoggingStatusForKey = Object.freeze({
     code: 'NoLoggingStatusForKey',
     message: 'There is no such thing as a logging status subresource for a key.',
     http_code: 400,
-}, {
+});
+S3Error.NoSuchBucket = Object.freeze({
     code: 'NoSuchBucket',
     message: 'The specified bucket does not exist.',
     http_code: 404,
-}, {
+});
+S3Error.NoSuchKey = Object.freeze({
     code: 'NoSuchKey',
     message: 'The specified key does not exist.',
     http_code: 404,
-}, {
+});
+S3Error.NoSuchLifecycleConfiguration = Object.freeze({
     code: 'NoSuchLifecycleConfiguration',
     message: 'The lifecycle configuration does not exist.',
     http_code: 404,
-}, {
+});
+S3Error.NoSuchUpload = Object.freeze({
     code: 'NoSuchUpload',
     message: 'The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.',
     http_code: 404,
-}, {
+});
+S3Error.NoSuchVersion = Object.freeze({
     code: 'NoSuchVersion',
     message: 'Indicates that the version ID specified in the request does not match an existing version.',
     http_code: 404,
-}, {
+});
+S3Error.NotImplemented = Object.freeze({
     code: 'NotImplemented',
     message: 'A header you provided implies functionality that is not implemented.',
     http_code: 501,
-}, {
+});
+S3Error.NotSignedUp = Object.freeze({
     code: 'NotSignedUp',
     message: 'Your account is not signed up for the Amazon S3 service. You must sign up before you can use, Amazon S3. You can sign up at the following URL: http://aws.amazon.com/s3',
     http_code: 403,
-}, {
+});
+S3Error.NoSuchBucketPolicy = Object.freeze({
     code: 'NoSuchBucketPolicy',
     message: 'The specified bucket does not have a bucket policy.',
     http_code: 404,
-}, {
+});
+S3Error.OperationAborted = Object.freeze({
     code: 'OperationAborted',
     message: 'A conflicting conditional operation is currently in progress against this resource. Try again.',
     http_code: 409,
-}, {
+});
+S3Error.PermanentRedirect = Object.freeze({
     code: 'PermanentRedirect',
     message: 'The bucket you are attempting to access must be addressed using the specified endpoint. Send all future requests to this endpoint.',
     http_code: 301,
-}, {
+});
+S3Error.PreconditionFailed = Object.freeze({
     code: 'PreconditionFailed',
     message: 'At least one of the preconditions you specified did not hold.',
     http_code: 412,
-}, {
+});
+S3Error.Redirect = Object.freeze({
     code: 'Redirect',
     message: 'Temporary redirect.',
     http_code: 307,
-}, {
+});
+S3Error.RestoreAlreadyInProgress = Object.freeze({
     code: 'RestoreAlreadyInProgress',
     message: 'Object restore is already in progress.',
     http_code: 409,
-}, {
+});
+S3Error.RequestIsNotMultiPartContent = Object.freeze({
     code: 'RequestIsNotMultiPartContent',
     message: 'Bucket POST must be of the enclosure-type multipart/form-data.',
     http_code: 400,
-}, {
+});
+S3Error.RequestTimeout = Object.freeze({
     code: 'RequestTimeout',
     message: 'Your socket connection to the server was not read from or written to within the timeout period.',
     http_code: 400,
-}, {
+});
+S3Error.RequestTimeTooSkewed = Object.freeze({
     code: 'RequestTimeTooSkewed',
     message: 'The difference between the request time and the server\'s time is too large.',
     http_code: 403,
-}, {
+});
+S3Error.RequestTorrentOfBucketError = Object.freeze({
     code: 'RequestTorrentOfBucketError',
     message: 'Requesting the torrent file of a bucket is not permitted.',
     http_code: 400,
-}, {
+});
+S3Error.SignatureDoesNotMatch = Object.freeze({
     code: 'SignatureDoesNotMatch',
     message: 'The request signature we calculated does not match the signature you provided. Check your AWS secret access key and signing method. For more information, see REST Authentication and SOAP Authentication for details.',
     http_code: 403,
-}, {
+});
+S3Error.ServiceUnavailable = Object.freeze({
     code: 'ServiceUnavailable',
     message: 'Reduce your request rate.',
     http_code: 503,
-}, {
+});
+S3Error.SlowDown = Object.freeze({
     code: 'SlowDown',
     message: 'Reduce your request rate.',
     http_code: 503,
-}, {
+});
+S3Error.TemporaryRedirect = Object.freeze({
     code: 'TemporaryRedirect',
     message: 'You are being redirected to the bucket while DNS updates.',
     http_code: 307,
-}, {
+});
+S3Error.TokenRefreshRequired = Object.freeze({
     code: 'TokenRefreshRequired',
     message: 'The provided token must be refreshed.',
     http_code: 400,
-}, {
+});
+S3Error.TooManyBuckets = Object.freeze({
     code: 'TooManyBuckets',
     message: 'You have attempted to create more buckets than allowed.',
     http_code: 400,
-}, {
+});
+S3Error.UnexpectedContent = Object.freeze({
     code: 'UnexpectedContent',
     message: 'This request does not support content.',
     http_code: 400,
-}, {
+});
+S3Error.UnresolvableGrantByEmailAddress = Object.freeze({
     code: 'UnresolvableGrantByEmailAddress',
     message: 'The email address you provided does not match any account on record.',
     http_code: 400,
-}, {
+});
+S3Error.UserKeyMustBeSpecified = Object.freeze({
     code: 'UserKeyMustBeSpecified',
     message: 'The bucket POST must contain the specified field name. If it is specified, check the order of the fields.',
     http_code: 400,
-}, {
-    // not defined in AWS list, but defined in the protocol handling
+});
+
+
+
+/////////////////////////////////////
+// Errors for generic HTTP replies //
+/////////////////////////////////////
+S3Error.NotModified = Object.freeze({
     code: 'NotModified',
     message: 'The resource was not modified according to the conditions in the provided headers.',
     http_code: 304,
-}, {
-    // not defined in AWS list
+});
+S3Error.BadRequest = Object.freeze({
     code: 'BadRequest',
     message: 'Bad Request',
     http_code: 400,
-}, {
-    // not defined in AWS list
-    // needed for error cases that require no code in the reply
-    reply_without_code: true,
-    code: 'BadRequestWithoutCode',
+});
+S3Error.BadRequestWithoutCode = Object.freeze({
+    // same as BadRequest but without encoding the <Code> field
+    // was needed for one of the cases in ceph/s3tests
     message: 'Bad Request',
     http_code: 400,
-}, {
-    // not defined in AWS docs but this is what they return
+});
+
+////////////////////////////////////////////////////////////////
+// Errors actually returned by AWS S3 although not documented //
+////////////////////////////////////////////////////////////////
+S3Error.ReplicationConfigurationNotFoundError = Object.freeze({
     code: 'ReplicationConfigurationNotFoundError',
     message: 'The replication configuration was not found',
     http_code: 404,
-}, {
-    // not defined in AWS docs but this is what they return
+});
+S3Error.NoSuchWebsiteConfiguration = Object.freeze({
     code: 'NoSuchWebsiteConfiguration',
     message: 'The specified bucket does not have a website configuration',
     http_code: 404,
-}, {
-    // not defined in AWS docs but this is what they return
+});
+S3Error.XAmzContentSHA256Mismatch = Object.freeze({
     code: 'XAmzContentSHA256Mismatch',
     message: 'The provided \'x-amz-content-sha256\' header does not match what was computed.',
     http_code: 400,
     // ClientComputedContentSHA256: '...',
     // S3ComputedContentSHA256: '...',
-}, {
-    // not defined in AWS docs but this is what they return
+});
+S3Error.MalformedPolicy = Object.freeze({
     code: 'MalformedPolicy',
     message: 'Invalid principal in policy',
     http_code: 400,
-    detail: '...',
-}, {
-    // not defined in AWS docs but this is what they return
+    detail: '...', // will be overridden from rpc_data, see handle_error in s3_rest.js
+});
+S3Error.NoSuchObjectLockConfiguration = Object.freeze({
     code: 'NoSuchObjectLockConfiguration',
     message: 'The specified object does not have a ObjectLock configuration',
     http_code: 404,
-}, {
-    // not defined in AWS docs but this is what they return
+});
+S3Error.ObjectLockConfigurationNotFoundError = Object.freeze({
     code: 'ObjectLockConfigurationNotFoundError',
     message: 'Object Lock configuration does not exist for this bucket',
     http_code: 404,
-
-}];
-
-for (const err_def of errors_defs) {
-    S3Error[err_def.code] = err_def;
-}
+});
 
 exports.S3Error = S3Error;

--- a/src/endpoint/s3/s3_rest.js
+++ b/src/endpoint/s3/s3_rest.js
@@ -2,8 +2,6 @@
 'use strict';
 
 const _ = require('lodash');
-const fs = require('fs');
-const path = require('path');
 
 const dbg = require('../../util/debug_module')(__filename);
 const config = require('../../../config');
@@ -310,6 +308,7 @@ async function authorize_request_policy(req) {
     if (!req.params.bucket) return;
     if (req.op_name === 'put_bucket') return;
     const policy_info = await req.object_sdk.read_bucket_sdk_policy_info(req.params.bucket);
+    if (!policy_info) return;
     const s3_policy = policy_info.s3_policy;
     const system_owner = policy_info.system_owner;
     const bucket_owner = policy_info.bucket_owner;
@@ -408,8 +407,8 @@ function get_bucket_and_key(req) {
 
 function parse_op_name(req) {
     const method = req.method.toLowerCase();
+    const { bucket, key, is_virtual_hosted_bucket } = get_bucket_and_key(req);
 
-    let { bucket, key, is_virtual_hosted_bucket } = get_bucket_and_key(req);
     req.params = { bucket, key };
     if (is_virtual_hosted_bucket) {
         req.virtual_hosted_bucket = bucket;
@@ -420,19 +419,23 @@ function parse_op_name(req) {
         return `${method}_service`;
     }
 
+    const query_keys = Object.keys(req.query);
+
     // bucket url
     if (!key) {
-        const query_keys = Object.keys(req.query);
         for (let i = 0; i < query_keys.length; ++i) {
-            if (BUCKET_SUB_RESOURCES[query_keys[i]]) return `${method}_bucket_${BUCKET_SUB_RESOURCES[query_keys[i]]}`;
+            if (BUCKET_SUB_RESOURCES[query_keys[i]]) {
+                return `${method}_bucket_${BUCKET_SUB_RESOURCES[query_keys[i]]}`;
+            }
         }
         return `${method}_bucket`;
     }
 
     // object url
-    const query_keys = Object.keys(req.query);
     for (let i = 0; i < query_keys.length; ++i) {
-        if (OBJECT_SUB_RESOURCES[query_keys[i]]) return `${method}_object_${OBJECT_SUB_RESOURCES[query_keys[i]]}`;
+        if (OBJECT_SUB_RESOURCES[query_keys[i]]) {
+            return `${method}_object_${OBJECT_SUB_RESOURCES[query_keys[i]]}`;
+        }
     }
     return `${method}_object`;
 }
@@ -599,15 +602,81 @@ function consume_usage_report() {
 }
 
 function load_ops() {
-    return js_utils.deep_freeze(new js_utils.PackedObject(
-        _.mapValues(
-            _.mapKeys(
-                fs.readdirSync(path.join(__dirname, 'ops')),
-                file => file.match(/^s3_([a-zA-Z0-9_]+)\.js$/)[1]
-            ),
-            file => require(`./ops/${file}`) // eslint-disable-line global-require
-        )
-    ));
+    /* eslint-disable global-require */
+    return js_utils.deep_freeze({
+        delete_bucket: require('./ops/s3_delete_bucket'),
+        delete_bucket_analytics: require('./ops/s3_delete_bucket_analytics'),
+        delete_bucket_cors: require('./ops/s3_delete_bucket_cors'),
+        delete_bucket_encryption: require('./ops/s3_delete_bucket_encryption'),
+        delete_bucket_inventory: require('./ops/s3_delete_bucket_inventory'),
+        delete_bucket_lifecycle: require('./ops/s3_delete_bucket_lifecycle'),
+        delete_bucket_metrics: require('./ops/s3_delete_bucket_metrics'),
+        delete_bucket_policy: require('./ops/s3_delete_bucket_policy'),
+        delete_bucket_replication: require('./ops/s3_delete_bucket_replication'),
+        delete_bucket_tagging: require('./ops/s3_delete_bucket_tagging'),
+        delete_bucket_website: require('./ops/s3_delete_bucket_website'),
+        delete_object: require('./ops/s3_delete_object'),
+        delete_object_tagging: require('./ops/s3_delete_object_tagging'),
+        delete_object_uploadId: require('./ops/s3_delete_object_uploadId'),
+        get_bucket: require('./ops/s3_get_bucket'),
+        get_bucket_accelerate: require('./ops/s3_get_bucket_accelerate'),
+        get_bucket_acl: require('./ops/s3_get_bucket_acl'),
+        get_bucket_analytics: require('./ops/s3_get_bucket_analytics'),
+        get_bucket_cors: require('./ops/s3_get_bucket_cors'),
+        get_bucket_encryption: require('./ops/s3_get_bucket_encryption'),
+        get_bucket_inventory: require('./ops/s3_get_bucket_inventory'),
+        get_bucket_lifecycle: require('./ops/s3_get_bucket_lifecycle'),
+        get_bucket_location: require('./ops/s3_get_bucket_location'),
+        get_bucket_logging: require('./ops/s3_get_bucket_logging'),
+        get_bucket_metrics: require('./ops/s3_get_bucket_metrics'),
+        get_bucket_notification: require('./ops/s3_get_bucket_notification'),
+        get_bucket_object_lock: require('./ops/s3_get_bucket_object_lock'),
+        get_bucket_policy: require('./ops/s3_get_bucket_policy'),
+        get_bucket_replication: require('./ops/s3_get_bucket_replication'),
+        get_bucket_requestPayment: require('./ops/s3_get_bucket_requestPayment'),
+        get_bucket_tagging: require('./ops/s3_get_bucket_tagging'),
+        get_bucket_uploads: require('./ops/s3_get_bucket_uploads'),
+        get_bucket_versioning: require('./ops/s3_get_bucket_versioning'),
+        get_bucket_versions: require('./ops/s3_get_bucket_versions'),
+        get_bucket_website: require('./ops/s3_get_bucket_website'),
+        get_object: require('./ops/s3_get_object'),
+        get_object_acl: require('./ops/s3_get_object_acl'),
+        get_object_legal_hold: require('./ops/s3_get_object_legal_hold'),
+        get_object_retention: require('./ops/s3_get_object_retention'),
+        get_object_tagging: require('./ops/s3_get_object_tagging'),
+        get_object_uploadId: require('./ops/s3_get_object_uploadId'),
+        get_service: require('./ops/s3_get_service'),
+        head_bucket: require('./ops/s3_head_bucket'),
+        head_object: require('./ops/s3_head_object'),
+        post_bucket: require('./ops/s3_post_bucket'),
+        post_bucket_delete: require('./ops/s3_post_bucket_delete'),
+        post_object_uploadId: require('./ops/s3_post_object_uploadId'),
+        post_object_uploads: require('./ops/s3_post_object_uploads'),
+        put_bucket: require('./ops/s3_put_bucket'),
+        put_bucket_accelerate: require('./ops/s3_put_bucket_accelerate'),
+        put_bucket_acl: require('./ops/s3_put_bucket_acl'),
+        put_bucket_analytics: require('./ops/s3_put_bucket_analytics'),
+        put_bucket_cors: require('./ops/s3_put_bucket_cors'),
+        put_bucket_encryption: require('./ops/s3_put_bucket_encryption'),
+        put_bucket_inventory: require('./ops/s3_put_bucket_inventory'),
+        put_bucket_lifecycle: require('./ops/s3_put_bucket_lifecycle'),
+        put_bucket_logging: require('./ops/s3_put_bucket_logging'),
+        put_bucket_metrics: require('./ops/s3_put_bucket_metrics'),
+        put_bucket_notification: require('./ops/s3_put_bucket_notification'),
+        put_bucket_object_lock: require('./ops/s3_put_bucket_object_lock'),
+        put_bucket_policy: require('./ops/s3_put_bucket_policy'),
+        put_bucket_replication: require('./ops/s3_put_bucket_replication'),
+        put_bucket_requestPayment: require('./ops/s3_put_bucket_requestPayment'),
+        put_bucket_tagging: require('./ops/s3_put_bucket_tagging'),
+        put_bucket_versioning: require('./ops/s3_put_bucket_versioning'),
+        put_bucket_website: require('./ops/s3_put_bucket_website'),
+        put_object: require('./ops/s3_put_object'),
+        put_object_acl: require('./ops/s3_put_object_acl'),
+        put_object_legal_hold: require('./ops/s3_put_object_legal_hold'),
+        put_object_retention: require('./ops/s3_put_object_retention'),
+        put_object_tagging: require('./ops/s3_put_object_tagging'),
+        put_object_uploadId: require('./ops/s3_put_object_uploadId'),
+    });
 }
 
 // EXPORTS

--- a/src/test/unit_tests/test_nb_native_hashes.js
+++ b/src/test/unit_tests/test_nb_native_hashes.js
@@ -8,14 +8,14 @@ const nb_native = require('../../util/nb_native');
 
 mocha.describe('nb_native hashes', function() {
     function md5(input) {
-        const MD5 = new(nb_native().MD5_MB)();
+        const MD5 = new (nb_native().MD5_MB)();
         const native_md5 = MD5.update(Buffer.from(input)).digest().toString('hex');
         const crypto_md5 = crypto.createHash('md5').update(input).digest('hex');
         assert.strictEqual(native_md5, crypto_md5);
     }
 
     function sha1(input) {
-        const SHA1 = new(nb_native().SHA1_MB)();
+        const SHA1 = new (nb_native().SHA1_MB)();
         const native_sha1 = SHA1.update(Buffer.from(input)).digest().toString('hex');
         const crypto_sha1 = crypto.createHash('sha1').update(input).digest('hex');
         assert.strictEqual(native_sha1, crypto_sha1);

--- a/src/util/fips.js
+++ b/src/util/fips.js
@@ -55,10 +55,11 @@ function set_fips_mode(mode = detect_fips_mode()) {
     nb_native().set_fips_mode(mode);
     if (mode) {
         // monkey-patch the crypto.createHash() function to provide a non-crypto md5 flow
+        const { MD5_MB } = nb_native();
         crypto.createHash = function(algorithm, options) {
             switch (algorithm) {
                 case 'md5': {
-                    return new HashWrap(new(nb_native().MD5_MB)());
+                    return new HashWrap(new MD5_MB());
                 }
                 default:
                     return original_crypto.createHash(algorithm, options);

--- a/src/util/js_utils.js
+++ b/src/util/js_utils.js
@@ -77,6 +77,11 @@ function named_array_push(obj, arr_name, item) {
     return arr;
 }
 
+/**
+ * @template T
+ * @param {T} obj
+ * @returns {T}
+ */
 function deep_freeze(obj) {
 
     // Checking isFrozen is the break condition for the recursion

